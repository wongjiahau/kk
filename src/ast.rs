use crate::non_empty::NonEmpty;
/// The syntax tree here represents raw syntax tree that is not type checked

#[derive(Debug, Clone)]
pub enum Statement {
    Let {
        keyword_let: Token,
        left: Token,
        type_variables: Vec<Token>,
        right: Expression,
        type_annotation: Option<TypeAnnotation>,
    },
    Type {
        keyword_type: Token,
        left: Token,
        right: TypeAnnotation,
        type_variables: Vec<Token>,
    },
    Enum {
        keyword_enum: Token,
        name: Token,
        constructors: Vec<EnumConstructor>,
        type_variables: Vec<Token>,
    },
    Do {
        keyword_do: Token,
        expression: Expression,
    },
}

#[derive(Debug, Clone)]
pub struct EnumConstructor {
    pub name: Token,
    pub left_parenthesis: Token,
    pub payload: Option<Box<TypeAnnotation>>,
    pub right_parenthesis: Token,
}

#[derive(Debug, Clone)]
pub struct TypeVariable {
    pub token: Token,
}

#[derive(Debug, Clone)]
pub enum Type {
    Underscore,

    /// Type variable that is declared. Cannot be substituted before instantiation.
    /// Note that a type variable is only explicit within its own scope
    ExplicitTypeVariable {
        name: String,
    },

    /// Type variable that is implicitly created  for unification.
    ImplicitTypeVariable {
        name: String,
    },
    Record {
        key_type_pairs: Vec<(String, Type)>,
    },
    Named {
        name: String,
        type_arguments: Vec<(String, Type)>,
    },
    Function(FunctionType),
    Tuple(Box<NonEmpty<Type>>),
    Boolean,
    Float,
    Integer,
    String,
    Character,
    Null,
    Array(Box<Type>),
}

#[derive(Debug, Clone)]
pub struct TagType {
    pub tagname: String,
    pub payload: Option<Box<Type>>,
}

#[derive(Debug, Clone)]
pub struct TypeScheme {
    pub type_variables: Vec<String>,
    pub type_value: Type,
}

#[derive(Debug, Clone)]
pub struct FunctionType {
    pub parameters_types: Box<NonEmpty<Type>>,
    pub return_type: Box<Type>,
}

#[derive(Debug, Clone)]
pub enum SymbolSource {
    BuiltIn,
    UserDefined {
        filename: String,
        position: Position,
    },
}

#[derive(Debug, Clone)]
pub struct NamedTypeAnnotationArguments {
    pub left_angular_bracket: Token,
    pub arguments: Vec<TypeAnnotation>,
    pub right_angular_bracket: Token,
}

#[derive(Debug, Clone)]
pub enum TypeAnnotation {
    Named {
        name: Token,
        type_arguments: Option<TypeArguments>,
    },
    Record {
        left_curly_bracket: Token,
        key_type_annotation_pairs: Vec<(Token, TypeAnnotation)>,
        right_curly_bracket: Token,
    },
    Array {
        left_square_bracket: Token,
        element_type: Box<TypeAnnotation>,
        right_square_bracket: Token,
    },
    Underscore(Token),
    Function {
        start_token: Token,
        parameters_types: Box<NonEmpty<TypeAnnotation>>,
        return_type: Box<TypeAnnotation>,
    },
}

#[derive(Debug, Clone)]
pub enum DestructurePattern {
    String(Token),
    Character(Token),
    Integer(Token),
    Boolean {
        token: Token,
        value: bool,
    },
    Null(Token),
    Underscore(Token),
    Identifier(Token),
    EnumConstructor {
        scoped_name: ScopedName,
        left_parenthesis: Token,
        payload: Option<Box<DestructurePattern>>,
        right_parenthesis: Token,
    },
    Record {
        left_curly_bracket: Token,
        key_value_pairs: Vec<DestructuredRecordKeyValue>,
        right_curly_bracket: Token,
    },
    Array {
        left_square_bracket: Token,
        right_square_bracket: Token,
        spread: Option<DestructurePatternArraySpread>,
    },
    Tuple(DestructurePatternTuple),
}
#[derive(Debug, Clone)]
pub struct DestructurePatternTuple {
    pub parentheses: Option<(/*left*/ Token, /*right*/ Token)>,
    pub values: Box<NonEmpty<DestructurePattern>>,
}

#[derive(Debug, Clone)]
pub struct DestructurePatternArraySpread {
    pub first_element: Box<DestructurePattern>,
    pub spread_token: Token,
    pub rest_elements: Box<DestructurePattern>,
}

#[derive(Debug, Clone)]
pub struct DestructuredRecordKeyValue {
    pub key: Token,
    pub type_annotation: Option<TypeAnnotation>,
    pub as_value: Option<DestructurePattern>,
}

#[derive(Debug, Clone)]
pub enum Expression {
    Null(Token),
    Boolean {
        token: Token,
        value: bool,
    },
    Float(Token),
    Integer(Token),
    String(Token),
    Character(Token),
    Variable(Token),
    EnumConstructor {
        scoped_name: ScopedName,
        left_parenthesis: Token,
        payload: Option<Box<Expression>>,
        right_parenthesis: Token,
    },
    Function(Box<Function>),
    FunctionCall(Box<FunctionCall>),
    Record {
        left_curly_bracket: Token,
        spread: Option<Box<Expression>>,
        key_value_pairs: Vec<RecordKeyValue>,
        right_curly_bracket: Token,
    },
    RecordAccess {
        expression: Box<Expression>,
        property_name: Token,
    },
    Array {
        left_square_bracket: Token,
        elements: Vec<Expression>,
        right_square_bracket: Token,
    },
    Let {
        keyword_let: Token,
        left: Box<DestructurePattern>,
        type_annotation: Option<TypeAnnotation>,
        right: Box<Expression>,
        true_branch: Box<Expression>,
        false_branch: Option<Box<Function>>,
    },
}

#[derive(Debug, Clone)]
pub struct RecordKeyValue {
    pub key: Token,
    pub type_annotation: Option<TypeAnnotation>,
    pub value: Expression,
}

#[derive(Debug, Clone)]
pub struct FunctionCall {
    pub function_name: Token,
    pub first_argument: Box<Expression>,
    pub rest_arguments: Option<FunctionCallRestArguments>,
    pub type_arguments: Option<TypeArguments>,
}

#[derive(Debug, Clone)]
pub struct TypeArguments {
    pub left_angular_bracket: Token,
    pub substitutions: Vec<(/*type_variable_name*/ Token, TypeAnnotation)>,
    pub right_angular_bracket: Token,
}

#[derive(Debug, Clone)]
pub struct FunctionCallRestArguments {
    pub left_parenthesis: Token,
    pub arguments: Vec<Expression>,
    pub right_parenthesis: Token,
}

#[derive(Debug, Clone)]
pub struct Function {
    pub branches: NonEmpty<FunctionBranch>,
}

#[derive(Debug, Clone)]
pub struct FunctionBranch {
    pub start_token: Token,
    pub parameters: FunctionParameters,
    pub body: Box<Expression>,
    pub return_type_annotation: Option<TypeAnnotation>,
}

impl FunctionBranch {
    pub fn parameters(&self) -> NonEmpty<FunctionParameter> {
        match &self.parameters {
            FunctionParameters::NoParenthesis { parameter } => NonEmpty {
                head: parameter.clone(),
                tail: vec![],
            },
            FunctionParameters::WithParenthesis { parameters, .. } => parameters.clone(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum FunctionParameters {
    NoParenthesis {
        parameter: FunctionParameter,
    },
    WithParenthesis {
        left_parenthesis: Token,
        parameters: NonEmpty<FunctionParameter>,
        right_parenthesis: Token,
    },
}

#[derive(Debug, Clone)]
pub struct FunctionBranchRestArguments {
    pub left_parenthesis: Token,
    pub rest_arguments: Vec<FunctionParameter>,
    pub right_parenthesis: Token,
}

#[derive(Debug, Clone)]
pub struct FunctionParameter {
    pub destructure_pattern: DestructurePattern,
    pub type_annotation: Option<TypeAnnotation>,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub position: Position,
    pub representation: String,
}

#[derive(Debug, Clone)]
pub struct ScopedName {
    pub namespaces: Vec<Token>,
    pub name: Token,
}

impl Token {
    pub fn dummy_identifier(representation: String) -> Token {
        Token {
            token_type: TokenType::Identifier,
            position: Position {
                line_start: 0,
                line_end: 0,
                character_index_start: 0,
                character_index_end: 0,
                column_start: 0,
                column_end: 0,
            },
            representation,
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenType {
    KeywordLet,
    KeywordType,
    KeywordEnum,
    KeywordDo,
    KeywordElse,
    KeywordNull,
    KeywordTrue,
    KeywordFalse,
    Whitespace,
    LeftCurlyBracket,
    RightCurlyBracket,
    LeftParenthesis,
    RightParenthesis,
    LeftSquareBracket,
    RightSquareBracket,
    Newline,
    Colon,
    LessThan,
    MoreThan,
    Equals,
    Period,
    Spread,
    Comma,
    Minus,
    FatArrowRight,
    ThinArrowRight,
    Backslash,
    Underscore,
    Identifier,
    String,
    Character,
    Integer,
    Float,
    ScopeResolution,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Source {
    File { path: String },
    NonFile { env_name: String },
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub struct Position {
    pub line_start: usize,
    pub line_end: usize,
    pub column_start: usize,
    pub column_end: usize,
    pub character_index_start: usize,
    pub character_index_end: usize,
}
