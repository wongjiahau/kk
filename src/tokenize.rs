use std::iter::Peekable;
use std::vec::IntoIter;

use crate::non_empty::NonEmpty;
use crate::parse::{ParseErrorKind, Parser};
use crate::{parse::ParseError, raw_ast::*};
use peeking_take_while::PeekableExt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Character {
    pub index: usize,
    pub line_number: usize,
    pub column_number: usize,
    pub value: char,
}

impl Character {
    pub fn position(&self) -> Position {
        Position {
            line_start: self.line_number,
            line_end: self.line_number,
            column_start: self.column_number,
            column_end: self.column_number,
            character_index_start: self.index,
            character_index_end: self.index,
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum TokenizeError {
    CharacterLiteralCannotBeEmpty {
        position: Position,
    },
    UnterminatedCharacterLiteral {
        position: Position,
    },
    UnterminatedMultilineComment {
        position: Position,
    },
    UnterminatedStringLiteral {
        position: Position,
    },
    InvalidToken {
        error: String,
        position: Position,
    },
    UnexpectedCharacter {
        position: Position,
        expected_character_value: char,
    },
    UnexpectedEof {
        expected_character_value: char,
    },
}

impl TokenizeError {
    pub fn into_parse_error(self) -> ParseError {
        ParseError {
            context: None,
            kind: ParseErrorKind::TokenizeError(self),
        }
    }
}

pub struct Tokenizer {
    characters_iterator: Peekable<IntoIter<Character>>,
    peeked_tokens: Vec<Token>,
}

impl Tokenizer {
    pub fn new(input: String) -> Tokenizer {
        let characters: Vec<Character> = input
            .split('\n')
            .enumerate()
            .flat_map(|(line_number, line)| {
                let mut characters = line.chars().into_iter().collect::<Vec<char>>();
                characters.push('\n');
                characters
                    .into_iter()
                    .enumerate()
                    .map(move |(column_number, character)| (character, line_number, column_number))
                    .collect::<Vec<(char, usize, usize)>>()
            })
            .enumerate()
            .map(
                |(index, (character, line_number, column_number))| Character {
                    index,
                    value: character,
                    line_number,
                    column_number,
                },
            )
            .into_iter()
            .collect::<Vec<Character>>();

        Tokenizer::from_character_iter(characters.into_iter())
    }

    pub fn from_character_iter(characters: IntoIter<Character>) -> Tokenizer {
        Tokenizer {
            characters_iterator: characters.peekable(),
            peeked_tokens: vec![],
        }
    }

    pub fn remaining_characters(self) -> Vec<Character> {
        self.characters_iterator.collect()
    }

    fn eat_character(&mut self, value: char) -> Result<Character, ParseError> {
        match self.characters_iterator.next() {
            Some(character) => {
                if character.value == value {
                    Ok(character)
                } else {
                    Err(TokenizeError::UnexpectedCharacter {
                        position: make_position(character, None),
                        expected_character_value: value,
                    }
                    .into_parse_error())
                }
            }
            None => Err(TokenizeError::UnexpectedEof {
                expected_character_value: value,
            }
            .into_parse_error()),
        }
    }

    pub fn peek(&mut self) -> Result<Option<Token>, ParseError> {
        match self.peeked_tokens.last() {
            Some(token) => Ok(Some(token.clone())),
            None => match self.next() {
                Ok(result) => match result {
                    Some(token) => {
                        self.peeked_tokens.push(token.clone());
                        Ok(Some(token))
                    }
                    None => Ok(None),
                },
                Err(error) => Err(error),
            },
        }
    }

    pub fn next(&mut self) -> Result<Option<Token>, ParseError> {
        // Clear peeked token
        if let Some(token) = self.peeked_tokens.pop() {
            return Ok(Some(token));
        }

        if let Some(character) = self.characters_iterator.next() {
            match character.value {
                // Comments
                '/' => match self.characters_iterator.next() {
                    Some(Character { value: '/', .. }) => {
                        let comment = self
                            .characters_iterator
                            .by_ref()
                            .peeking_take_while(|character| character.value != '\n')
                            .collect::<Vec<Character>>();

                        Ok(Some(Token {
                            token_type: TokenType::Comment,
                            position: make_position(character, comment.last()),
                            representation: stringify(comment),
                        }))
                    }
                    Some(Character { value: '*', .. }) => {
                        let mut comment = vec![character.clone()];
                        let mut found_asterisk = false;
                        let comment = loop {
                            match self.characters_iterator.next() {
                                Some(character) => {
                                    if found_asterisk && character.value == '/' {
                                        comment.push(character);
                                        break comment;
                                    }
                                    found_asterisk = character.value == '*';
                                    comment.push(character);
                                }
                                None => {
                                    return Err(ParseError {
                                        context: None,
                                        kind: ParseErrorKind::TokenizeError(
                                            TokenizeError::UnterminatedMultilineComment {
                                                position: make_position(character, comment.last()),
                                            },
                                        ),
                                    })
                                }
                            }
                        };
                        Ok(Some(Token {
                            position: make_position(character, comment.last()),
                            token_type: TokenType::MultilineComment {
                                characters: comment.clone(),
                            },
                            representation: stringify(comment),
                        }))
                    }
                    Some(other) => Err(ParseError {
                        context: None,
                        kind: ParseErrorKind::TokenizeError(TokenizeError::UnexpectedCharacter {
                            position: make_position(other, None),
                            expected_character_value: '/',
                        }),
                    }),
                    None => Err(ParseError {
                        context: None,
                        kind: ParseErrorKind::TokenizeError(TokenizeError::UnexpectedEof {
                            expected_character_value: '/',
                        }),
                    }),
                },
                // Character
                '\'' => match self.characters_iterator.next() {
                    Some(quote @ Character { value: '\'', .. }) => {
                        Err(TokenizeError::CharacterLiteralCannotBeEmpty {
                            position: make_position(character, Some(&quote)),
                        }
                        .into_parse_error())
                    }
                    Some(c) => match self.characters_iterator.next() {
                        Some(end_quote @ Character { value: '\'', .. }) => Ok(Some(Token {
                            token_type: TokenType::Character,
                            representation: format!("'{}'", c.value),
                            position: make_position(character, Some(&end_quote)),
                        })),
                        other => Err(TokenizeError::UnterminatedCharacterLiteral {
                            position: make_position(character, other.as_ref()),
                        }
                        .into_parse_error()),
                    },
                    None => Err(TokenizeError::UnterminatedCharacterLiteral {
                        position: make_position(character, None),
                    }
                    .into_parse_error()),
                },
                // Tag
                '`' => {
                    let mut characters = self
                        .characters_iterator
                        .by_ref()
                        .peeking_take_while(|character| character.value != '`')
                        .collect::<Vec<Character>>();

                    match self.characters_iterator.next() {
                        Some(backtick) => {
                            characters.push(backtick);

                            let representation =
                                format!("{}{}", character.value, stringify(characters.clone()));
                            Ok(Some(Token {
                                // token_type:  get_token_type(representation.clone()),
                                token_type: TokenType::Tag,
                                representation,
                                position: make_position(character, characters.last()),
                            }))
                        }
                        None => panic!("missing closing backtick(`) for tag"),
                    }
                }
                // String
                '"' => {
                    let start_quote = character;
                    enum StartOf {
                        Nothing,
                        Escape,
                        StringInterpolation,
                    }
                    let mut interpolated_string_sections: Vec<InterpolatedStringSection> = vec![];
                    let (end_quote, characters) = {
                        let mut characters: Vec<Character> = vec![];
                        let mut start_of = StartOf::Nothing;
                        let end_quote = loop {
                            if let Some(character) = self.characters_iterator.next() {
                                let (character, next_start_of) = match &character.value {
                                    '\\' => (
                                        Some(character),
                                        match start_of {
                                            StartOf::Escape => StartOf::Nothing,
                                            _ => StartOf::Escape,
                                        },
                                    ),
                                    '"' => match start_of {
                                        StartOf::Escape => (Some(character), StartOf::Nothing),
                                        _ => break Ok(character),
                                    },
                                    '$' => (None, StartOf::StringInterpolation),
                                    '{' => {
                                        if let StartOf::StringInterpolation = start_of {
                                            interpolated_string_sections.push(
                                                InterpolatedStringSection::String(stringify(
                                                    characters.clone(),
                                                )),
                                            );
                                            characters.clear();
                                            let mut parser = Parser::new(self);
                                            interpolated_string_sections.push(
                                                InterpolatedStringSection::Expression(Box::new(
                                                    parser.parse_expression()?,
                                                )),
                                            );
                                            parser.eat_token(TokenType::RightCurlyBracket, None)?;
                                            (None, StartOf::Nothing)
                                        } else {
                                            (Some(character), StartOf::Nothing)
                                        }
                                    }
                                    _ => (Some(character), StartOf::Nothing),
                                };

                                if let Some(character) = character {
                                    characters.push(character.clone());
                                }
                                start_of = next_start_of;
                            } else {
                                break Err(TokenizeError::UnterminatedStringLiteral {
                                    position: make_position(start_quote.clone(), characters.last()),
                                }
                                .into_parse_error());
                            }
                        }?;
                        (end_quote, characters)
                    };

                    match interpolated_string_sections.split_first() {
                        Some((head, interpolated_string_sections)) => {
                            let interpolated_string_sections = {
                                let mut interpolated_string_sections =
                                    interpolated_string_sections.to_vec();
                                interpolated_string_sections
                                    .push(InterpolatedStringSection::String(stringify(characters)));
                                interpolated_string_sections
                            };
                            Ok(Some(Token {
                                token_type: TokenType::InterpolatedString {
                                    start_quote: Box::new(start_quote.clone()),
                                    sections: NonEmpty {
                                        head: head.clone(),
                                        tail: interpolated_string_sections,
                                    },
                                    end_quote: Box::new(end_quote.clone()),
                                },
                                representation: "".to_string(),
                                position: make_position(start_quote, Some(&end_quote)),
                            }))
                        }
                        None => Ok(Some(Token {
                            token_type: TokenType::String,
                            representation: format!("\"{}\"", stringify(characters)),
                            position: make_position(start_quote, Some(&end_quote)),
                        })),
                    }
                }
                // Number
                '-' | '0'..='9' => {
                    let intergral = self
                        .characters_iterator
                        .by_ref()
                        .peeking_take_while(|character| character.value.is_digit(10))
                        .collect::<Vec<Character>>();

                    if character.value == '-' && intergral.is_empty() {
                        let characters = self
                            .characters_iterator
                            .by_ref()
                            .peeking_take_while(|character| is_symbol(character.value))
                            .collect::<Vec<Character>>();

                        let representation =
                            format!("{}{}", character.value, stringify(characters.clone()));
                        return Ok(Some(Token {
                            token_type: TokenType::Identifier,
                            representation,
                            position: make_position(character, characters.last()),
                        }));
                    };

                    match self.characters_iterator.peek() {
                        Some(Character { value: '.', .. }) => {
                            let period = self.characters_iterator.next().unwrap();
                            let fractional = self
                                .characters_iterator
                                .peeking_take_while(|character| character.value.is_digit(10))
                                .collect::<Vec<Character>>();

                            if fractional.is_empty() {
                                // means there's no fractional part
                                // that means the next token is a Period
                                self.peeked_tokens.push(Token {
                                    token_type: TokenType::Period,
                                    representation: period.value.to_string(),
                                    position: Position {
                                        line_start: period.line_number,
                                        line_end: period.line_number,
                                        column_start: period.column_number,
                                        column_end: period.column_number,
                                        character_index_start: period.index,
                                        character_index_end: period.index,
                                    },
                                });
                                Ok(Some(Token {
                                    token_type: TokenType::Integer,
                                    representation: format!(
                                        "{}{}",
                                        character.value,
                                        stringify(intergral.clone())
                                    ),
                                    position: make_position(character, intergral.last()),
                                }))
                            } else {
                                Ok(Some(Token {
                                    token_type: TokenType::Float,
                                    representation: format!(
                                        "{}{}.{}",
                                        character.value,
                                        stringify(intergral),
                                        stringify(fractional.clone())
                                    ),
                                    position: make_position(character, fractional.last()),
                                }))
                            }
                        }
                        _ => Ok(Some(Token {
                            token_type: TokenType::Integer,
                            representation: format!(
                                "{}{}",
                                character.value,
                                stringify(intergral.clone())
                            ),
                            position: make_position(character, intergral.last()),
                        })),
                    }
                }
                // Identifier
                '#' | 'A'..='Z' | 'a'..='z' => {
                    let characters = self
                        .characters_iterator
                        .by_ref()
                        .peeking_take_while(|character| {
                            character.value.is_alphanumeric() || character.value == '_'
                        })
                        .collect::<Vec<Character>>();

                    let representation =
                        format!("{}{}", character.value, stringify(characters.clone()));
                    Ok(Some(Token {
                        // token_type:  get_token_type(representation.clone()),
                        token_type: TokenType::Identifier,
                        representation,
                        position: make_position(character, characters.last()),
                    }))
                }
                ':' => Ok(Some(Token {
                    token_type: TokenType::Colon,
                    representation: ":".to_string(),
                    position: make_position(character, None),
                })),
                '.' => Ok(Some(Token {
                    token_type: TokenType::Period,
                    representation: ".".to_string(),
                    position: make_position(character, None),
                })),
                // Symbolic identifer
                other if is_symbol(other) => {
                    let characters = self
                        .characters_iterator
                        .by_ref()
                        .peeking_take_while(|character| is_symbol(character.value))
                        .collect::<Vec<Character>>();

                    let representation =
                        format!("{}{}", character.value, stringify(characters.clone()));
                    Ok(Some(Token {
                        token_type: TokenType::Identifier,
                        representation,
                        position: make_position(character, characters.last()),
                    }))
                }
                other => Ok(Some(Token {
                    position: make_position(character, None),
                    representation: other.to_string(),
                    token_type: match other {
                        '{' => TokenType::LeftCurlyBracket,
                        '}' => TokenType::RightCurlyBracket,
                        '(' => TokenType::LeftParenthesis,
                        ')' => TokenType::RightParenthesis,
                        '[' => TokenType::LeftSquareBracket,
                        ']' => TokenType::RightSquareBracket,
                        ' ' => TokenType::Whitespace,
                        ',' => TokenType::Comma,
                        '\n' => TokenType::Newline,
                        other => TokenType::Other(other),
                    },
                })),
            }
        } else {
            Ok(None)
        }
    }
}

fn is_symbol(c: char) -> bool {
    "!@#$%^&*_-<>=+/|?~".contains(c)
}

pub fn make_position(first: Character, last: Option<&Character>) -> Position {
    let last = match last {
        Some(character) => character,
        None => &first,
    };
    Position {
        column_start: first.column_number,
        column_end: last.column_number,
        line_start: first.line_number,
        line_end: last.line_number,
        character_index_start: first.index,
        character_index_end: last.index,
    }
}

pub fn stringify(characters: Vec<Character>) -> String {
    characters
        .into_iter()
        .map(|character| character.value.to_string())
        .collect::<Vec<String>>()
        .join("")
}

pub fn get_token_type(s: String) -> TokenType {
    if s.eq("with") {
        TokenType::KeywordWith
    } else if s.eq("switch") {
        TokenType::KeywordSwitch
    } else if s.eq("case") {
        TokenType::KeywordCase
    } else if s.eq("if") {
        TokenType::KeywordIf
    } else if s.eq("else") {
        TokenType::KeywordElse
    } else if s.eq("let") {
        TokenType::KeywordLet
    } else if s.eq("type") {
        TokenType::KeywordType
    } else if s.eq("enum") {
        TokenType::KeywordEnum
    } else if s.eq("do") {
        TokenType::KeywordDo
    } else if s.eq("true") {
        TokenType::KeywordTrue
    } else if s.eq("false") {
        TokenType::KeywordFalse
    } else if s.eq("null") {
        TokenType::KeywordNull
    } else if s.eq("import") {
        TokenType::KeywordImport
    } else if s.eq("from") {
        TokenType::KeywordFrom
    } else if s.eq("as") {
        TokenType::KeywordAs
    } else if s.eq("export") {
        TokenType::KeywordExport
    } else if s.eq("interface") {
        TokenType::KeywordInterface
    } else if s.eq("implements") {
        TokenType::KeywordImplements
    } else if s.eq("where") {
        TokenType::KeywordWhere
    } else {
        TokenType::Identifier
    }
}
