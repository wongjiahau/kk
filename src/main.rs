pub mod ast;
use ast::*;

mod transpile;
use transpile::*;

mod parse;
use parse::*;

mod tokenize;
use tokenize::*;

fn main() {
    println!(
        "{:?}",
        transpile_source("let f = (x: number, y) => 'helo' | x => '2'".to_string())
    )
}

pub fn transpile_source(source: String) -> Result<String, ParseError> {
    let tokens = tokenize(source)?;
    let statements = parse_statements(tokens)?;
    // println!("{:#?}", statements);
    Ok(transpile_statements(statements))
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_tokenize_arrow_right() {
        assert_eq!(
            tokenize("=>".to_string()),
            Ok(vec![Token {
                token_type: TokenType::ArrowRight,
                representation: "=>".to_string(),
                position: Position {
                    column_start: 0,
                    column_end: 1,
                    line_start: 0,
                    line_end: 0
                }
            }])
        );
    }

    #[test]
    fn test_tokenize_tag() {
        assert_eq!(
            tokenize("#Hello_world123 ".to_string()),
            Ok(vec![
                Token {
                    token_type: TokenType::Tag("#Hello_world123".to_string()),
                    representation: "#Hello_world123".to_string(),
                    position: Position {
                        column_start: 0,
                        column_end: 14,
                        line_start: 0,
                        line_end: 0
                    }
                },
                Token {
                    token_type: TokenType::Whitespace,
                    representation: " ".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 15,
                        column_end: 15,
                    }
                }
            ])
        );
    }

    #[test]
    fn test_tokenize_string() {
        let string = "'qwertyuiop \"world\" 1234567890!'".to_string();
        assert_eq!(
            tokenize(string.clone()),
            Ok(vec![Token {
                token_type: TokenType::String(string.clone()),
                representation: string.clone(),
                position: Position {
                    column_start: 0,
                    column_end: string.len() - 1,
                    line_start: 0,
                    line_end: 0
                }
            }])
        )
    }

    #[test]
    fn test_tokenize_identifier() {
        let string = "Hello_world_123".to_string();
        assert_eq!(
            tokenize(string.clone()),
            Ok(vec![Token {
                token_type: TokenType::Identifier(string.clone()),
                representation: string.clone(),
                position: Position {
                    column_start: 0,
                    column_end: string.len() - 1,
                    line_start: 0,
                    line_end: 0
                }
            }])
        )
    }

    #[test]
    fn test_tokenize_keyword() {
        assert_eq!(
            tokenize("let type".to_string()),
            Ok(vec![
                Token {
                    token_type: TokenType::KeywordLet,
                    representation: "let".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 0,
                        column_end: 2
                    }
                },
                Token {
                    token_type: TokenType::Whitespace,
                    representation: " ".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 3,
                        column_end: 3
                    }
                },
                Token {
                    token_type: TokenType::KeywordType,
                    representation: "type".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 4,
                        column_end: 7
                    }
                }
            ])
        )
    }

    #[test]
    fn test_let_statement_1() {
        assert_eq!(
            transpile_source("let x = '123'".to_string()),
            Ok("const x = '123'".to_string())
        )
    }

    #[test]
    fn test_let_statement_2() {
        assert_eq!(
            transpile_source("let x = y".to_string()),
            Ok("const x = y".to_string())
        )
    }

    #[test]
    fn test_function_1() {
        assert_debug_snapshot!(transpile_source("let x = \\x => 'yo'".to_string()))
    }
    #[test]
    fn test_function_2() {
        assert_debug_snapshot!(transpile_source(
            "let x = \\#red => 'red' \\ #blue => 'blue'".to_string()
        ))
    }

    #[test]
    fn test_function_3() {
        assert_debug_snapshot!(transpile_source(
            "let x = \\#ok(#some(r)) => r \\ _ => 'yo'".to_string()
        ))
    }

    #[test]
    fn test_function_4() {
        assert_debug_snapshot!(transpile_source(
            "let or = \\#false, #false => #false \\ _, _ => #true".to_string()
        ))
    }

    #[test]
    fn test_function_5() {
        assert_debug_snapshot!(transpile_source("let square = \\x:number => x".to_string()))
    }

    #[test]
    fn test_function_call_1() {
        assert_debug_snapshot!(transpile_source("let x = y.square.bomb".to_string()))
    }

    #[test]
    fn test_function_call_2() {
        assert_debug_snapshot!(transpile_source("let x = a.times(b).and(c, d)".to_string()))
    }

    #[test]
    fn test_function_call_3() {
        assert_debug_snapshot!(transpile_source(
            "let x = x.(\\#ok(x) => a \\#no => b)".to_string()
        ))
    }

    #[test]
    fn test_function_call_4() {
        assert_debug_snapshot!(transpile_source(
            "let x = 'hello world'.replace('hello', with: 'bye')".to_string()
        ))
    }
}
