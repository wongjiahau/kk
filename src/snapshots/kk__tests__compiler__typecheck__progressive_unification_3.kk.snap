---
source: src/main.rs
expression: stripped_actual_output.trim()
---
============
INPUT
============
let f
  : \number => null
  = \(x): number => 2

============
EXIT CODE
============
1

============
STDOUT
============


============
STDERR
============
error:
  ┌─ tests/compiler/typecheck/progressive_unification_3.kk:3:11
  │
3 │   = \(x): number => 2
  │           ^^^^^^ Wrong type annotation
  │
  = The expected type annotation is:

      null
